### DISCUSSION TITLE
🔬 방법론 #1: PC-Router — 예측 코딩 기반 적응적 추론 라우터

### DISCUSSION THREAD

**Discussion by hgKwak** [hgKwak/Brain-cognition-inspired-AI#33] [Research Topics]
# 🔬 PC-Router: Predictive Coding 기반 적응적 추론 라우터

> **Parent**: [#34](https://github.com/hgKwak/Brain-cognition-inspired-AI/discussions/34) (전체 로드맵) | **RQ1** | **Priority: P0**

---

## 1. 동기 (Motivation)

뇌의 예측 코딩(Predictive Coding)은 상위 레벨이 하위 레벨의 입력을 예측하고, **예측 오차(prediction error)가 클수록 더 많은 인지 자원을 할당**하는 메커니즘이다 (Rao & Ballard, 1999; Friston, 2005).

현재 LLM 추론 라우팅의 한계:
- **CogER** (Hu et al., 2025): 4단계 추론을 제안하나, 라우팅이 RL 기반이라 뇌과학적 근거 부족
- **Ziabari et al. (2025)**: 엔트로피 기반 하이브리드가 최적이라 보였으나, 엔트로피만으로는 쿼리의 의미적 복잡도를 충분히 포착하지 못함
- **Dualformer** (Su et al., 2024): 단일 모델 내 모드 전환이나, control token이 수동적

**핵심 가설**: 예측 오차(PE)를 쿼리 복잡도의 프록시로 활용하면, 엔트로피보다 풍부한 의미적 신호를 제공하여 더 정확한 추론 경로 라우팅이 가능하다.

---

### 2.1 Predictive Coding Router 아키텍처

```
Input query x
    │
    ▼
┌────────────────────┐
│ LLM Encoder         │  → hidden states h₁, h₂, ..., hₙ
└────────┬───────────┘
         │
         ▼
┌────────────────────┐
│ Top-Down Predictor  │  → predicted hidden states ĥ₁, ĥ₂, ..., ĥₙ
│ (learned generative │     (상위 레벨이 하위 레벨 예측)
│  model)             │
└────────┬───────────┘
         │
         ▼
┌────────────────────┐
│ Prediction Error    │  → PE = f(h, ĥ) 
│ Computation         │     (layer-wise prediction error)
└────────┬───────────┘
         │
         ▼
┌────────────────────┐
│ Complexity Estimator│  → c = g(PE₁, PE₂, ..., PEₗ)
│ (MLP head)          │     (aggregated complexity score)
└────────┬───────────┘
         │
    ┌────┼────┬────┐
    L1   L2   L3   L4
```

### 2.2 핵심 수식

**Prediction Error 계산:**
```
PE_l = ||h_l - ĥ_l||₂  (layer l에서의 예측 오차)
```

**Top-Down Predictor:**
```
ĥ_l = W_l · h_{l+1} + b_l  (상위 레이어에서 하위 레이어 예측)
```

**Complexity Score:**
```
c = σ(W_c · [PE_1; PE_2; ...; PE_L] + b_c)  ∈ [0, 1]
```

**라우팅 결정:**
```
Level = {
  L1 (Direct Answer)     if c < τ₁
  L2 (Chain-of-Thought)  if τ₁ ≤ c < τ₂
  L3 (Deep Reasoning)    if τ₂ ≤ c < τ₃
  L4 (Tool-Augmented)    if c ≥ τ₃
}
```

### 2.3 학습 방법

**Phase 1: Top-Down Predictor 사전학습**
- Self-supervised: LLM의 인접 레이어 간 prediction을 학습
- Loss: MSE(h_l, ĥ_l)

**Phase 2: Router 학습**
- 라벨: 각 쿼리에 대해 oracle complexity level (4단계) 부여
  - 전략: 모든 쿼리를 L1~L4로 풀어보고, 최소 레벨에서 정답을 맞춘 레벨을 oracle label로 사용
- Loss: Cross-entropy + 효율성 페널티 (α · level_cost)
- RL fine-tuning (optional): GRPO로 비용-정확도 트레이드오프 최적화

### 2.4 비교 대상 라우터 (Baselines)

| 라우터 | 신호 | 출처 |
|--------|------|------|
| Entropy-based | 출력 토큰 엔트로피 | Ziabari et al. (2025) |
| RL-based (CogER) | MDP 정책 | Hu et al. (2025) |
| Prompt-length | 입력 길이 | Heuristic |
| Random | 균등 분포 | Baseline |
| **PC-Router (Ours)** | **예측 오차** | **제안** |

---

### 3.1 필요 리소스
- Base LLM: Qwen2.5-7B 또는 Llama3-8B (오픈소스)
- Top-Down Predictor: 경량 MLP (레이어당 ~1M params)
- Complexity Estimator: 소형 분류기 (~0.5M params)
- GPU: A100 80GB × 4 (학습), A100 × 1 (추론)

### 3.2 구현 단계
- [ ] Step 1: Base LLM에서 layer-wise hidden states 추출 파이프라인 구축
- [ ] Step 2: Top-Down Predictor 학습 (self-supervised)
- [ ] Step 3: Oracle complexity label 데이터셋 구축
- [ ] Step 4: Router (Complexity Estimator) 학습
- [ ] Step 5: 4단계 추론 경로별 프롬프트/디코딩 전략 구현
- [ ] Step 6: End-to-end 통합 및 inference pipeline 구축

---

## 4. 검증 실험 (문서 [#30](https://github.com/hgKwak/Brain-cognition-inspired-AI/discussions/30)에서 상세 설계)

- 라우팅 정확도: Oracle level 대비 일치율
- 추론 정확도: 벤치마크별 정확도
- 연산 효율: 평균 토큰 수, FLOPs, 지연시간
- 라우팅 해석가능성: PE 패턴 시각화, 복잡도 vs PE 상관관계

---

## 관련 문헌
- Rao & Ballard (1999): Predictive Coding in the Visual Cortex
- Friston (2005): A Theory of Cortical Responses (Free Energy Principle)
- CogER (Hu et al., 2025): Cognitive Elastic Reasoning — [#35](https://github.com/hgKwak/Brain-cognition-inspired-AI/discussions/35) Evidence 8
- Ziabari et al. (2025): Reasoning on a Spectrum — [#35](https://github.com/hgKwak/Brain-cognition-inspired-AI/discussions/35) Evidence 2
- Dualformer (Su et al., 2024): Fast/Slow 모드 전환 — [#35](https://github.com/hgKwak/Brain-cognition-inspired-AI/discussions/35) Evidence 5
